### PR TITLE
MoshiAdapter support for AnyMessage

### PR DESCRIPTION
**Pull request by oldergod**
part of #1387

### CODE REVIEW

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      writer.nullValue()
+      return
+    }
+    writer.beginObject()
+
+    writer.name("@type")
+    writer.value(value.typeUrl)
+
+    val protoAdapter = typeUrlToAdapter[value.typeUrl]
+        ?: throw IOException("Cannot find type for url: ${value.typeUrl}")
```
Comment (swankjesse) [inline wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: JsonDataException

Comment (swankjesse) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: Also include `writer.getPath()` so we get a path in the exception message

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    val delegate = moshi.adapter(protoAdapter.type!!.java) as JsonAdapter<Message<*, *>>
+
+    val flattenToken = writer.beginFlatten()
+    delegate.toJson(writer, value.unpackOrNull(protoAdapter) as Message<*, *>)
```
Comment (swankjesse) [inline wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: why `orNull` if we’re immediately going to cast to non-null?

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    if (reader.peek() == JsonReader.Token.NULL) {
+      reader.nextNull<Any>()
+      return null
+    }
+
+    val typeUrl = reader.peekJson().use { it.readStringNamed("@type") }
+        ?: throw JsonDataException("expected @type in $reader")
+
+    val protoAdapter = typeUrlToAdapter[typeUrl]
+        ?: throw IOException("Cannot resolve type: $typeUrl")
```
Comment (swankjesse) [inline wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: JsonDataException

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    val delegate = moshi.adapter(protoAdapter.type!!.java) as JsonAdapter<Message<*, *>>
+
+    val value = delegate.fromJson(reader)
+
+    return AnyMessage.pack(value!!)
```
Comment (swankjesse) [inline wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: should this method take the adapter?

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/WireJsonAdapterFactory.kt [hunk 5]
```diff
@@ -43,7 +43,25 @@ import java.util.ArrayList
...
+class WireJsonAdapterFactory private constructor(
+  private val typeUrlToAdapter: Map<String, ProtoAdapter<*>>
+) : JsonAdapter.Factory {
+  constructor() : this(mapOf())
+
+  /**
+   * Returns a new WireJsonAdapterFactory that can encode the messages for [adapters] if they're
+   * used with [AnyMessage].
+   */
+  fun plus(vararg adapters: ProtoAdapter<*>) : WireJsonAdapterFactory {
```
Comment (swankjesse) [inline wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/WireJsonAdapterFactory.kt]: I think I also want an overload that takes a list? varargs methods don’t compose well

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt [hunk 6]
```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
...
+        .usingTypeRegistry(typeRegistry)
+    val protocMessageJson = jsonPrinter.print(protocMessage)
+
+    val wireMessage = PizzaDelivery(
+        address = "507 Cross Street",
+        pizzas = listOf(Pizza(toppings = listOf("pineapple", "onion"))),
+        promotion = AnyMessage.pack(BuyOneGetOnePromotion(coupon = "MAUI"))
+    )
+
+    val moshi = Moshi.Builder()
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: worth elevating these to be fields?\

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt [hunk 7]
```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
...
+        |    "@type": "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion",
+        |    "coupon": "MAUI"
+        |  }
+        |}
+        """.trimMargin()
+
+    try {
+      jsonAdapter.fromJson(json)
+      fail()
+    } catch (expected: IOException) {
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: JsonDataException

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt [hunk 8]
```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
...
+      assertThat(expected).hasMessage("Cannot resolve type: " +
+          "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion")
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: include the path in the expected message

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt [hunk 9]
```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
...
+
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory())
+        .build()
+
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+    try {
+      jsonAdapter.toJson(pizzaDelivery)
+      fail()
+    } catch (expected: AssertionError) {
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: What? This is definitely the wrong type to be throwing

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt [hunk 10]
```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
...
+        |  }
+        |}
+        """.trimMargin()
+
+    try {
+      jsonAdapter.fromJson(json)
+      fail()
+    } catch (expected: JsonDataException) {
+      assertThat(expected).hasMessage(
+          """expected @type in JsonReader([text=\n    "coupon": "MAUI"\n  }\n}])""")
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: I think I misspoke when I said include the JsonReader, instead I want `jsonReader.getPath()`

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt [hunk 11]
```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
...
+    }
+  }
+
+  // Nice error on unrecognized @type
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: Yes please

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt [hunk 12]
```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
...
+  // Nice error on absent @type
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: this is done

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt [hunk 13]
```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
...
+  // Nice error on unregistered type
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: same as line 269

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 9 above)
Comment (JakeWharton) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: `fail` also throws this so you never would actually fail!

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/WireJsonAdapterFactory.kt (hunk 5 above)
Comment (oldergod) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/WireJsonAdapterFactory.kt]: went with
```kotlin
  fun plus(adapter: ProtoAdapter<*>): WireJsonAdapterFactory {
// and
  fun plus(adapters: List<ProtoAdapter<*>>): WireJsonAdapterFactory {
```

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 9 above)
Comment (oldergod) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: hmmhmmhhm
https://github.com/square/moshi/blob/master/moshi/src/main/java/com/squareup/moshi/JsonAdapter.java#L62

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 11 above)
Comment (oldergod) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: This is actually what I thought I was already doing in `unregisteredTypeOnReading` and `unregisteredTypeOnWriting`.

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 9 above)
Comment (oldergod) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: Right but `JsonDataException` is runtime so it doesn't get wrapped anymore. Cool.

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 10 above)
Comment (oldergod) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: fixed

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 8 above)
Comment (oldergod) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: fixed

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 7 above)
Comment (oldergod) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: fixed

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt (hunk 3 above)
Comment (oldergod) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: fixed

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt (hunk 1 above)
Comment (oldergod) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: fixed

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt (hunk 2 above)
Comment (oldergod) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: fixed

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt (hunk 4 above)
Comment (oldergod) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/AnyMessageJsonAdapter.kt]: I don't think so? What do we gain doing it?

Current code
```kotlin
    fun pack(message: Message<*, *>): AnyMessage {
      val typeUrl = message.adapter.typeUrl
          ?: error("recompile ${message::class} to use it with AnyMessage")
      return AnyMessage(typeUrl, message.encodeByteString())
    }
```

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 6 above)
Comment (oldergod) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: Hmm, how about I say no for now.

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt (hunk 8 above)
Comment (swankjesse) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt]: fixed?

**Comment by oldergod** [comment]
Fixed the last missing piece https://github.com/square/wire/compare/d2fdde99d371966b1545c3f46bbe70b497eca541..58e7eff879211a5ed0aca99d838e6199fd38e501

will merge on green.

**Pull request merged**